crates/publish/src/oci/client.rs

```diff
@@ -209,28 +214,89 @@ impl Client {
         ))
     }
 
+    /// Save a credential set containing the registry username and password.
+    pub async fn login(
+        server: impl AsRef<str>,
+        username: impl AsRef<str>,
+        password: impl AsRef<str>,
+    ) -> Result<()> {
+        // We want to allow a user to login to both https://ghcr.io and ghcr.io.
+        let server = match server.as_ref().parse::<Url>() {
+            Ok(url) => url
+                .host_str()
+                .unwrap_or_else(|| server.as_ref())
```

Comment:
nit: this is one of the few cases where aliasing variables is really nice: `let server = server.as_ref();`